research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -0,0 +1,607 @@
+        pytorch iterators
+
+        Args:
+            config (Config): The config file from the server
+
+        Returns:
+            Tuple[DataLoader, DataLoader]: A tuple of length two. The client
+                train and validation dataloaders as pytorch dataloaders
+        """
+        # flwr 1.9.0 now raises an error any time a process ends.
+        # To prevent errors due to this since dataloaders use multiprocessing
+        # lets ovverride that behaviour before the child processes are created
+        fl_sigint_handler = signal.getsignal(signal.SIGINT)
+        fl_sigterm_handler = signal.getsignal(signal.SIGTERM)
+        signal.signal(signal.SIGINT, og_sigint_handler)
+        signal.signal(signal.SIGTERM, og_sigterm_handler)
+
+        # Get the nnunet dataloader iterators
+        with nostdout():
+            train_loader, val_loader = self.nnunet_trainer.get_dataloaders()
+
+        # Set the signal handlers back to what they were for flwr
+        signal.signal(signal.SIGINT, fl_sigint_handler)
+        signal.signal(signal.SIGTERM, fl_sigterm_handler)
+
+        # The batchgenerators package used under the hood by the dataloaders
+        # creates an additional stream handler for the root logger
+        # Therefore all logs get printed twice, We can fix this by clearing the
+        # root logger handlers.
+        root_logger = logging.getLogger()
+        root_logger.handlers.clear()
+
+        # By default nnunet dataloaders are infinite, so if we are training by
+        # num epochs we need to make them not infinite anymore
+        if config["local_epochs"] is not None:
+            infinite = False
+
+        assert isinstance(self.nnunet_config, str)
+        assert is_valid_config(self.nnunet_config)
+        train_loader = nnUNetDLWrapper(
+            nnunet_dataloader=train_loader, nnunet_config=self.nnunet_config, infinite=infinite
+        )
+        val_loader = nnUNetDLWrapper(nnunet_dataloader=val_loader, nnunet_config=self.nnunet_config, infinite=infinite)
+
+        return train_loader, val_loader
+
+    def get_model(self, config: Config) -> nn.Module:
+        return self.nnunet_trainer.network
+
+    def get_criterion(self, config: Config) -> _Loss:
+        return Module2LossWrapper(self.nnunet_trainer.loss)
+
+    def get_optimizer(self, config: Config) -> Optimizer:
+        return self.nnunet_trainer.optimizer
+
+    def create_plans(self, config: Config) -> Dict[str, Any]:
+        """
+        Modifies the provided plans file to work with the local client dataset
+
+        Args:
+            config (Config): The config provided by the server. Expects the
+                'nnunet_plans' key with a pickled dictionary as the value
+
+        Returns:
+            Dict[str, Any]: The modified nnunet plans for the client
+        """
+        # Get the nnunet plans specified by the server
+        plans = pickle.loads(config["nnunet_plans"])  # type: ignore
```

Comment:
research/picai/fl_nnunet/nnunet_client.py:222: error: Argument 1 to "loads" has
incompatible type "Union[bool, bytes, float, int, str]"; expected "Buffer" 
[arg-type]
            plans = pickle.loads(config["nnunet_plans"])

Reply:
Ah, I think I see. I'm sort of surprised that the above works, given that the config entry that you're getting is a string object rather than a buffer. I might suggest a type safe way of doing this is
```
```